Add jni Support for API CreateColumnFamilyWithImport

mayuehappy opened the pull request:
- Add the following missing options to src/main/java/org/rocksdb/ImportColumnFamilyOptions.java and in java/rocksjni/import_column_family_options.cc in RocksJava.
- Add the struct to src/main/java/org/rocksdb/ExportImportFilesMetaData.java and in java/rocksjni/export_import_files_metadatajni.cc in RocksJava.
- Add New Java API `createColumnFamilyWithImport` to src/main/java/org/rocksdb/RocksDB.java 
- Add New Java API `exportColumnFamily` to src/main/java/org/rocksdb/Checkpoint.java


Test plan:
- added unit tests for exportColumnFamily in org.rocksdb.CheckpointTest
- added unit tests for createColumnFamilyWithImport to org.rocksdb.ImportColumnFamilyTest

adamretter commented on java/src/main/java/org/rocksdb/ExportImportFilesMetaData.java:
```diff
@@ -0,0 +1,56 @@
+  /**
+   * Called from JNI C++
+   */
+  public ExportImportFilesMetaData(final byte[] dbComparatorName, final LiveFileMetaData[] files) {
```
If this is only called from JNI, it can be made `private` please

adamretter commented on java/src/main/java/org/rocksdb/ExportImportFilesMetaData.java:
```diff
@@ -0,0 +1,56 @@
+    return Arrays.asList(files);
+  }
+
+  public long newExportImportFilesMetaDataHandle() {
```
I don't think this should be in this class. I see it is called from RocksDB. I think it would be better if the processing was done in RocksDB as the native handles are not used here.

mayuehappy commented on java/src/main/java/org/rocksdb/RocksDB.java:
```diff
@@ -749,6 +749,58 @@ public List<ColumnFamilyHandle> createColumnFamilies(
+      final List<ExportImportFilesMetaData> metadatas) throws RocksDBException {
+    final int metadataNum = metadatas.size();
+    final long[] metadataHandeList = new long[metadataNum];
+    for (int i = 0; i < metadataNum; i++) {
```
@adamretter 

Thank you very much for Review code, it is my honor to contribute code.

I‘v read your comment. Before I modify the code I would like to ask a question, PTAL

ExportImportFilesMetaData may be used in two places. One situation is that you can create an ExportImportFilesMetaData through Checkpoint.exportColumnFamily, from here it seems that rg.rocksdb.ExportImportFilesMetaData and rg.rocksdb.LiveFileMetaData are similar, Called from JNI C++, so i did not to follow that pattern.
But in another case, ExportImportFilesMetaData needs to be passed in when calling createColumnFamilyWithImport. In this case we need to follow the pattern of inheriting from RocksObject.
So if this is the case, do LiveFileMetaData and SstFileMetaData also need to follow this pattern?

adamretter replied:
1. So in C++ when you call `Checkpoint::ExportColumnFamily` the caller will be the owner of the new `ExportImportFilesMetaData` object that is created and returned. So if you are calling this from Java, then the Java caller is now the owner of this, and is responsible for freeing it - for this something like `RocksObject` in Java is perfect.

2. In C++ when you call `DB::CreateColumnFamilyWithImport` as the caller you have to provide an `ExportImportFilesMetaData` object. As you are in C++ you may choose to create this object either on the stack or on the heap. From Java though if you want to have a Java object which represents this C++ object, then you will need to call JNI and create the object on the heap, at which point you are the owner of that object, and so therefore again `RocksObject` in Java again becomes the correct choice.

@mayuehappy Is that helpful?

mayuehappy commented:
@adamretter @cbi42  Ping~ , Can anyone help me look at this PR?

adamretter commented:
@mayuehappy I just replied to your comment above - let me know if you need any more guidance...

mayuehappy commented:
@adamretter 
Thanks for the guidance, I've updated the PR. Change ExportImportFilesMetaData to extends RocksObject, please help and take a look

adamretter commented on java/src/main/java/org/rocksdb/ExportImportFilesMetaData.java:
```diff
@@ -0,0 +1,23 @@
+  private ExportImportFilesMetaData(final long nativeHandle) {
+    super(nativeHandle);
+    // We do not own the native object!
+    disOwnNativeHandle();
```
@mayuehappy can you explain why you are callng `disOwnNativeHandle` here please?

mayuehappy replied:
@adamretter Thanks for the reviewing, I thought about this place again. `org.rocksdb.ExportImportFilesMetaData` should indeed own `ROCKSDB_NAMESPACE::ExportImportFilesMetaData`'s nativeHandle. My initial idea was a little wrong, so I deleted these two lines of code.

pnowojski commented:
Hi @mayuehappy and @adamretter. Thank you both for your efforts so far on this and other related PRs! I'm eagerly waiting for this feature to be merged. Is there something that I could do to help pushing this PR over the line? If there are still some issues to be resolved, me or my colleagues would be happy to help.

mayuehappy commented:
> Hi @mayuehappy and @adamretter. Thank you both for your efforts so far on this and other related PRs! I'm eagerly waiting for this feature to be merged. Is there something that I could do to help pushing this PR over the line? If there are still some issues to be resolved, me or my colleagues would be happy to help.

@pnowojski Hi ~ Thank you for your attention. I alse hope this PR can be merged soon so that Flink can use this new feature to speed up state recovery!

@adamretter .If there is anything that needs to be modified in this PR, please comment and I will fix it as soon as possible.
cc @cbi42

adamretter commented:
@mayuehappy I think this looks good. I would like my colleague @alanpaxton to also give it a quick once over too. @alanpaxton could you just check the "ownership" semantics as discussed above, I am pretty certain we have them correct now - but a second pair of eyes is always valuable.

alanpaxton reviewed (commented):
@adamretter @mayuehappy I have looked at this w.r.t. "ownership" semantics, and it appears correct to me for that. I noticed a couple of small issues in passing that I commented..

mayuehappy commented:
@cbi42 @adamretter @alanpaxton 
Thanks a lot for reviewing the code, I've updated the PR as suggested by @alanpaxton . Mainly include the following changes：
1. fix metadataHandeList typo issues
2. Updated `ImportColumnFamilyOptions` class comment. And use try-with-resources in unit tests to create ImportColumnFamilyOptions
3. change `Java_org_rocksdb_Checkpoint_exportColumnFamily` returns types to jlong, then call new ExportImportFilesMetaData(metadata_native_) to create ExportImportFilesMetaData and delete portal.h related changes

alanpaxton commented on java/src/main/java/org/rocksdb/ExportImportFilesMetaData.java:
```diff
@@ -6,14 +6,11 @@
  */
 public class ExportImportFilesMetaData extends RocksObject {
-  private ExportImportFilesMetaData(final long nativeHandle) {
+  public ExportImportFilesMetaData(final long nativeHandle) {
```
Can this still be package (default) rather than public ?

mayuehappy replied:
@alanpaxton Thanks for the reminder, I have updated it

cbi42 commented on java/src/main/java/org/rocksdb/Checkpoint.java:
```diff
@@ -5,6 +5,8 @@
 
 package org.rocksdb;
 
+import java.util.*;
```
CI complains about this import: "Unused import 'java.util.*'"

mayuehappy replied:
Thanks,  fix it
